ion/bits.go

```diff
@@ -244,41 +243,89 @@ func appendTag(b []byte, code byte, len uint64) []byte {
 	return appendVarUint(b, len)
 }
 
-// timeLen pre-calculates the length, in bytes, of the given time value.
-func timeLen(offset int, utc time.Time) uint64 {
-	ret := varIntLen(int64(offset))
+// timestampLen pre-calculates the length, in bytes, of the given timestamp value.
+func timestampLen(offset int, utc Timestamp) uint64 {
+	var ret uint64
 
-	// Almost certainly two but let's be safe.
-	ret += varUintLen(uint64(utc.Year()))
+	if utc.kind == Unspecified {
+		ret = 1
+	} else {
+		ret = varIntLen(int64(offset))
+	}
+
+	if utc.precision >= Year {
+		// Almost certainly two but let's be safe.
+		ret += varUintLen(uint64(utc.dateTime.Year()))
+
+		// Month, day, hour, minute, and second are all guaranteed to be one byte.
+		if utc.precision >= Month {
+			ret++
 
-	// Month, day, hour, minute, and second are all guaranteed to be one byte.
-	ret += 5
+			if utc.precision >= Day {
```

Comment:
I think this deeply nested `if` tree would be simpler/faster as a `switch` statement.

Reply:
Addressed.